### ISSUE TITLE
DLR loss implementation in TF2

### ISSUE THREAD

**Issue by CNOCycle** [fra31/auto-attack#37]
Hi authors,

The DLR loss in TF2 may be incomplete. Misclassified cases are not be implemented:

https://github.com/fra31/auto-attack/blob/42189aab794672131f7f7584f15710e41f5c540d/autoattack/utils_tf2.py#L196-L202

I tried to implement DLR loss from the original paper's description in Section 4.1. Please verify the following code's correctness.
```python
def dir_loss(x, y, num_classes=10):

    # logit
    logit = x
    logit_max = tf.reduce_max(logit, axis=1)
    logit_sort = tf.sort(logit, axis=1)

    # onthot_y
    #argmax_y = tf.argmax(y, axis = 1)
    argmax_y = y
    y_onehot = tf.one_hot(argmax_y , num_classes, dtype=tf.float32)
    logit_y = tf.reduce_sum(y_onehot * logit, axis=1)

    # z_i
    cond = (logit_max == logit_y)
    z_i = tf.where(cond, logit_sort[:, -2], logit_sort[:, -1])

    # loss
    z_y = logit_y
    z_p1 =  logit_sort[:, -1]
    z_p3 = logit_sort[:, -3]

    loss = - (z_y - z_i) / (z_p1 - z_p3 + 1e-12)

    return loss
```

If the code is acceptable, I will open a new PR.

**Comment by fra31** [comment]
Hi,

thanks for the contribution! It looks good to me, feel free to open a PR.
As a minor comment, is `argmax_y = y` necessary or would `tf.one_hot(y, ...)` work as well?

**Comment by CNOCycle** [comment]
> As a minor comment, is argmax_y = y necessary or would tf.one_hot(y, ...) work as well?

It dependent on what shape of `y` in the code.

In my own code, the term `y` is a one-hot vector, whose shape is `[batch_size, n_class]`; the argument `y` in this project is scalar, whose shape is `[batch_size, 1]`.


The operation `tf.reduce_sum(y_onehot * logit, axis=1)` is to get the logit belongs to `y` class. Because `y_onehot` has been a one-hot vector already, the element-wise multiplication ensures that other classes are masked.  Next, we do the `reduce_sum` along `axis=1` and the variable's shape is from `[batch_size, n_class]` to `[batch_size, 1]` and its logit belongs to `y` class.
